text/0000-glimmer-components.md

```diff
@@ -0,0 +1,1022 @@
+  post: null,
+
+  type: readOnly('post.type'),
+
+  didInsertElement() {
+    if (this.post.type === 'image') {
+      setupImageOverlay(this.element.querySelector('img'));
+    }
+  }
+});
+```
+
+And here is an equivalent component written with the Glimmer component API:
+
+```hbs
+<!-- templates/components/post.hbs -->
+<section ...attributes role="region" type={{@post.type}} class="post {{@post.type}}">
+  {{#if (eq @post.type 'image')}}
+    <img {{did-render this.didInsertImage}} src={{@post.imageUrl}} title={{@post.imageTitle}} />
+  {{/if}}
+
+  {{@post.text}}
+</section>
+```
+```js
+// components/post.js
+export default class PostComponent extends GlimmerComponent {
+  didInsertImage(element) {
+    setupImageOverlay(element);
+  }
+}
+```
+
+Glimmer components eliminate many of the common paper cuts that cause confusion
+with classic components, and align more closely with modern template syntax and
+features.
+
+### Outer HTML Semantics
+
+The biggest change Glimmer components make is defaulting to outer HTML
+semantics - What You See Is What You Get, in templates. In the classic component
+API, components had a implicit wrapper element. Given this component template:
+
+```hbs
+Hello, world!
+```
+
+The output by _default_ would be:
+
+```html
+<!-- OUTPUT -->
+<div>
+  Hello, world!
+</div>
+```
+
+But we can't know that for sure unless we look at the component definition. If
+we do, we might see that the outer wrapping element is actually a `section`, and
+it has a `.hello-world` class:
+
+```js
+export default Component.extend({
+  tagName: 'section',
+  classNames: ['hello-world']
+});
+```
+```html
+<!-- OUTPUT -->
+<section class="hello-world">
+  Hello, world!
+</section>
+```
+
+This behavior means that the template for a component is missing crucial
+information and context. Even for the simplest component, users must check the
+class definition to know with certainty what the full template of the component
+is. And unlike bindings, there is no hint to the user that there may be
+something dynamic that they should check on - without advanced knowledge of
+Ember's APIs, there is no way of knowing about this behavior.
+
+By contrast, Glimmer components have no wrapping outer element. There is no way
+to define class names, class name bindings, attribute bindings, or any other DOM
+element values. The template is the single source of truth for the output of a
+component, and any dynamic values are explicitly stated in it.
+
+```hbs
+<!-- template.hbs -->
+<section class="hello-world">
+  Hello, world!
+</section>
+```
+
+```html
+<!-- OUTPUT -->
+<section class="hello-world">
+  Hello, world!
+</section>
+```
+
+We can immediately see that this is a static component with no bindings, no
+dynamic values, and no meaningful state. Even if there was a component
+definition, we know that it is not in any way affecting the output of this
+template (though it could be "side-effecting" and changing state elsewhere).
```

Comment:
Is this aside about "side-effecting" needed to understand the overall point about the template being static? Without an example or explanation, I feel like this could cause confusion.

Reply:
Just being (overly) thorough, I think we can remove it.